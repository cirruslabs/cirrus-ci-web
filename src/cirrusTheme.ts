import { atom, selector } from 'recoil';

import { grey, lightGreen, orange, red } from '@mui/material/colors';
import { ThemeOptions } from '@mui/material/styles';

import { localStorageEffect } from 'utils/recoil';

export const prefersDarkModeState = atom({
  key: 'CurrentlyPrefersDarkMode',
  default: false,
  effects_UNSTABLE: [localStorageEffect('CurrentlyPreferredTheme')],
});

export const muiThemeOptions = selector({
  key: 'muiThemeOptions',
  get: ({ get }) => {
    const prefersDarkMode = get(prefersDarkModeState);
    return prefersDarkMode ? muiDarkTheme : muiLightTheme;
  },
});

export const cirrusColorsState = selector({
  key: 'cirrusColorsState',
  get: ({ get }) => {
    const prefersDarkMode = get(prefersDarkModeState);
    return {
      undefined: prefersDarkMode ? grey['600'] : grey['400'],
    };
  },
});

export const cirrusThemeOptions = selector({
  key: 'cirrusThemeOptions',
  get: ({ get }) => {
    const prefersDarkMode = get(prefersDarkModeState);
    return prefersDarkMode ? cirrusDarkTheme : cirrusLightTheme;
  },
});

let cirrusBaseTheme: ThemeOptions = {
  typography: {
    fontFamily: 'Roboto, sans-serif',
  },
  shape: {
    borderRadius: 2,
  },
};

export let cirrusLightTheme: ThemeOptions = {
  ...cirrusBaseTheme,
  palette: {
    mode: 'light',
    primary: {
      main: grey['900'],
      dark: grey['900'],
      light: grey['50'],
      contrastText: grey['50'],
    },
    secondary: {
      main: grey['700'],
      dark: grey['700'],
      light: grey['50'],
      contrastText: grey['50'],
    },
    action: {
      hover: grey['200'],
    },
    success: {
      light: lightGreen['300'],
      main: lightGreen['500'],
      dark: lightGreen['700'],
    },
    error: {
      light: red['300'],
      main: red['500'],
      dark: red['700'],
    },
    warning: {
      light: orange['300'],
      main: orange['500'],
      dark: orange['700'],
    },
  },
  components: {
    MuiChip: {
      styleOverrides: {
        avatar: {
          marginLeft: 0,
          marginRight: 0,
          height: 32,
          width: 32,
        },
      },
    },
    MuiSkeleton: {
      styleOverrides: {
        root: {
          backgroundColor: grey['800'],
        },
      },
    },
    MuiTableCell: {
      styleOverrides: {
        body: {
          padding: 6, // it was changed from v4
        },
      },
    },
  },
};

export let cirrusDarkTheme: ThemeOptions = {
  ...cirrusBaseTheme,
  palette: {
    mode: 'dark',
    primary: {
      main: grey['900'],
      dark: grey['900'],
      light: grey['50'],
      contrastText: grey['50'],
    },
    secondary: {
      main: grey['700'],
      dark: grey['700'],
      light: grey['50'],
      contrastText: grey['50'],
    },
    success: {
      light: lightGreen['400'],
      main: lightGreen['600'],
      dark: lightGreen['800'],
    },
    error: {
      light: red['400'],
      main: red['600'],
      dark: red['800'],
    },
    warning: {
      light: orange['400'],
      main: orange['600'],
      dark: orange['800'],
    },
  },
  components: {
    MuiChip: {
      styleOverrides: {
        avatar: {
          marginLeft: 0,
          marginRight: 0,
          height: 32,
          width: 32,
        },
      },
    },
    MuiTableCell: {
      styleOverrides: {
        body: {
          padding: 6, // it was changed from v4
        },
      },
    },
    MuiTab: {
      styleOverrides: {
        root: {
          '&.Mui-selected': {
            color: grey['50'],
          },
        },
      },
    },
  },
};

export let muiLightTheme: ThemeOptions = {
  palette: cirrusLightTheme.palette,
};

export let muiDarkTheme: ThemeOptions = {
  palette: cirrusDarkTheme.palette,
};

export const cirrusOpenDrawerState = atom({
  key: 'CirrusOpenDrawer',
  default: false,
  effects_UNSTABLE: [localStorageEffect('CirrusOpenDrawer')],
});
