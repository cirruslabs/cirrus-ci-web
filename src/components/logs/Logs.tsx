import React from 'react';
import AnsiUp from 'ansi_up';
import { useLocation, useNavigate } from 'react-router-dom';
import classNames from 'classnames';
import { makeStyles } from '@mui/styles';
import queryString from 'query-string';

let ansiFormatter = new (AnsiUp as any)();
ansiFormatter.use_classes = true;

const useStyles = makeStyles(theme => {
  return {
    logContainer: {
      overflowY: 'hidden',
      minHeight: '70px',
      height: '100%',
      maxWidth: '100%',
      background: theme.palette.primary.dark,
      padding: 8,
    },
    logLine: {
      width: '100%',
      display: 'flex',
      flexWrap: 'wrap',
      margin: 0,
      color: theme.palette.primary.contrastText,
      '&:hover': {
        background: theme.palette.secondary.dark,
      },
      '&:focus': {
        outline: 0,
      },
      '&:empty': {
        height: theme.typography.fontSize,
      },
    },
    logLineHighlighted: {
      background: theme.palette.secondary.dark,
    },
  };
});

interface Props {
  logsName: string;
  logs: string;
}

function Logs(props: Props) {
  const navigate = useNavigate();
  const location = useLocation();
  let highLightedLineStart = NaN;
  let highLightedLineEnd = NaN;

  let search = queryString.parse(location.search);

  if (search.logs === props.logsName || search.command === props.logsName) {
    let [startLine, endLine] = location.hash.replace('#', '').split('-');
    if (!endLine) {
      endLine = startLine;
    }

    highLightedLineStart = parseInt(startLine.replace('L', ''), 10);
    highLightedLineEnd = parseInt(endLine.replace('L', ''), 10);
    let elementToFocus = document.getElementById(startLine);
    if (elementToFocus) {
      elementToFocus.focus();
    }
  }

  function selectLine(event, lineNumber) {
    let lineRange = `L${lineNumber}`;
    if (event.shiftKey) {
      lineRange = `L${highLightedLineStart}-L${lineNumber}`;
    }
    navigate({
      search: `?logs=${props.logsName}`,
      hash: `#${lineRange}`,
    });
  }

  let classes = useStyles();
  return (
    <div className={classes.logContainer}>
      {props.logs.split('\n').map((line, index) => (
        <div
          id={'L' + index}
          tabIndex={0} // to make it focusable
          key={index}
          className={classNames('log-line', classes.logLine, {
            [classes.logLineHighlighted]: highLightedLineStart <= index && index <= highLightedLineEnd,
          })}
          onClick={e => selectLine(e, index)}
          dangerouslySetInnerHTML={{ __html: ansiFormatter.ansi_to_html(line) }}
        />
      ))}
    </div>
  );
}

export default Logs;
