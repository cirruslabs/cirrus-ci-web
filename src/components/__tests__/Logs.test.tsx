import React from 'react';
import Logs from '../logs/Logs';
import renderer from 'react-test-renderer';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { ThemeProvider } from '@mui/styles';
import { createTheme } from '@mui/material/styles';
import { cirrusLightTheme } from '../../cirrusTheme';
import { Theme } from '@mui/material';

declare module '@mui/styles/defaultTheme' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface DefaultTheme extends Theme {}
}

it('renders correctly', () => {
  const exampleLog = `\
hello world
I am a log
I log things
`;

  const tree = renderer
    .create(
      <ThemeProvider theme={createTheme(cirrusLightTheme)}>
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<Logs logs={exampleLog} logsName="test" />} />
          </Routes>
        </BrowserRouter>
      </ThemeProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
