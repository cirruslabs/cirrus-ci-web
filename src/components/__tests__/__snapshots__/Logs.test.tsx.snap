// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="makeStyles-logContainer-1"
>
  <div
    className="log-line makeStyles-logLine-2"
    dangerouslySetInnerHTML={
      Object {
        "__html": "hello world",
      }
    }
    id="L0"
    onClick={[Function]}
    tabIndex={0}
  />
  <div
    className="log-line makeStyles-logLine-2"
    dangerouslySetInnerHTML={
      Object {
        "__html": "I am a log",
      }
    }
    id="L1"
    onClick={[Function]}
    tabIndex={0}
  />
  <div
    className="log-line makeStyles-logLine-2"
    dangerouslySetInnerHTML={
      Object {
        "__html": "I log things",
      }
    }
    id="L2"
    onClick={[Function]}
    tabIndex={0}
  />
  <div
    className="log-line makeStyles-logLine-2"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    id="L3"
    onClick={[Function]}
    tabIndex={0}
  />
</div>
`;
